pkg/scheduler/scheduler.go

```diff
@@ -385,17 +385,47 @@ func buildQueueingHintMap(es []framework.EnqueueExtensions) internalqueue.Queuei
 		// cannot be moved by any regular cluster event.
 		// So, we can just ignore such EventsToRegister here.
 
+		registerNodeAdded := false
+		registerNodeTaintUpdated := false
 		for _, event := range events {
 			fn := event.QueueingHintFn
 			if fn == nil || !utilfeature.DefaultFeatureGate.Enabled(features.SchedulerQueueingHints) {
 				fn = defaultQueueingHintFn
 			}
 
+			if event.Event.Resource == framework.Node {
+				if event.Event.ActionType&framework.Add != 0 {
+					registerNodeAdded = true
+				}
+				if event.Event.ActionType&framework.UpdateNodeTaint != 0 {
+					registerNodeTaintUpdated = true
+				}
+			}
+
 			queueingHintMap[event.Event] = append(queueingHintMap[event.Event], &internalqueue.QueueingHintFunction{
 				PluginName:     e.Name(),
 				QueueingHintFn: fn,
 			})
 		}
+		if registerNodeAdded && !registerNodeTaintUpdated {
+			// Temporally fix for the issue https://github.com/kubernetes/kubernetes/issues/109437
+			// NodeAdded QueueingHint isn't always called because of preCheck.
+			// It's definitely not something expected for plugin developers,
+			// and registering UpdateNodeTaint event is the only mitigation for now.
+			//
+			// So, here registers UpdateNodeTaint event for plugins that has NodeAdded event, but don't have UpdateNodeTaint event.
+			// It has a bad impact for the requeuing efficiency though, a lot better than some Pods being stuch in the
+			// unschedulable pod pool.
+			// This behavior will be removed when we remove the preCheck feature.
+			// See: https://github.com/kubernetes/kubernetes/issues/110175
+			queueingHintMap[framework.ClusterEvent{Resource: framework.Node, ActionType: framework.UpdateNodeTaint}] =
+				append(queueingHintMap[framework.ClusterEvent{Resource: framework.Node, ActionType: framework.UpdateNodeTaint}],
```

Comment:
Why is this specific to taints? what about a node that receives a node label update that impacts the nodeAffinity preCheck? Also, why doesn't the plugin explicitly add those instead of implicitly doing it here (which seems to me the right thing to do)?

Reply:
cc @alculquicondor 

> what about a node that receives a node label update that impacts the nodeAffinity preCheck

https://github.com/kubernetes/kubernetes/pull/122292#issuecomment-1862867359 changed this PR's direction from adding a general node update to adding a node taint event.
You're actually right. If we want a complete fix (covering all minor scenarios), we should register `Node/Update` event here, but it'd hurt the throughput by requeueing many Pods unnecessarily.

So, the point would be which scenario(s) is likely enough to deserve a fix. And, I believe [this scenario](https://github.com/kubernetes/kubernetes/pull/122292#issuecomment-1861794637) is, while others aren't.

Reply:
Ok, I see. I think the right solution is to add the taint/toleration plugin to the list of UnschedulablePlugins for that pod when it fails preCheck on node add event. We could change the function signature of PreCheck to return the plugin that failed the pod.

Reply:
We are discussing the "right" solution for this PR. Since PreChecks continue to exist, then having them integrate with failedPlugins is certainly the better approach than this hack which could be surprising to plugins and is not really covering all corner cases.

I didn't get a chance to look at the implementation, but I didn't expect it to be that complicated.